src/Core/ServiceProvider.php

```diff
@@ -112,6 +113,10 @@ public function register(): void
      */
     public function provides(string $id): bool
     {
+        if (!$this->provides) {
```

Comment:
I don't think a method specifically meant to check the existence of something should be throwing an exception. This had the potential to cause problems.

Reply:
I do agree however that there is a logic problem with provides() not having the initial data to even make the check make sense.
Maybe this should go into 5.x next, though.

Reply:
Looking at https://container.thephpleague.com/4.x/service-providers/ and https://github.com/thephpleague/container/blob/4.x/src/ServiceProvider/ServiceProviderAggregate.php#L67-L76 I am also not sure if this is correct.

The reason service providers exist in the first place is to allow lazy loading services only when they are needed.
Therefore the container needs to know which services each provider can offer without actually initializing those objects.

The `services()` method is called only after the container has checked if the provider has the requested service definition inside it.

Reply:
Looking at the implementation again an empty $services property would almost always indicate a user error since it would return false for every provides() call.

Reply:
Jep, sure sounds like it would be a silent no-op, and as such not really useful to have.
So maybe this can help users indicate that there is sth wrong and needs to be fixed up (or this check not being used as it would never be able to return true).

Reply:
I would just move this to 5.next instead of adding t exception in a patch release.

Reply:
> Looking at the implementation again an empty $services property would almost always indicate a user error since it would return false for every provides() call.

Isn't this developer error though? This exception could help save people time figuring out why cake is 'broken'.

Reply:
Yes, that's what I meant.